src/cell_list.h

```diff
@@ -0,0 +1,214 @@
+            load();
+            current = sorted_elements[current_distance].begin();
+        }
+        std::pair<point, T> &operator*() {
+            assert( range );
+            return *current;
+        }
+        std::pair<point, T> *operator->() {
+            assert( range );
+            return &*current;
+        }
+
+        cell_list_iterator &operator++() {
+            assert( range );
+            ++current;
+            while( current == sorted_elements[current_distance].end() &&
+                   current_distance <= range->distance ) {
+                ++current_distance;
+                if( current_distance > distance_to_nearest_cell_wall ) {
+                    load();
+                }
+                current = sorted_elements[current_distance].begin();
+            }
+            if( current_distance > range->distance ) {
+                range = nullptr;
+            }
+            return *this;
+        }
+
+        bool operator==( const cell_list_iterator &rhs ) const {
+            if( range == nullptr && rhs.range == nullptr ) {
+                return true;
+            }
+            if( range != rhs.range ) {
+                return false;
+            }
+            return current == rhs.current;
+        }
+        bool operator!=( const cell_list_iterator &rhs ) const {
+            return !( *this == rhs );
+        }
+
+    private:
+        void add( const std::pair<point, T> &candidate ) {
+            const int distance = rl_dist( candidate.first, range->origin );
+            if( distance > range->distance ) {
+                return;
+            }
+            sorted_elements[distance].push_back( candidate );
+        }
+        void load() {
+            int min_dist = cell_ranks_loaded * CELL_SIZE + range->offset_distance;
+            min_dist = std::min( min_dist, range->distance );
+            assert( min_dist >= 0 );
+            distance_to_nearest_cell_wall = min_dist;
+            if( cell_ranks_loaded == 0 ) {
+                // Special case for first cell.
+                for( const std::pair<point, T> &candidate :
+                     range->list->cells[range->origin_cell.y][range->origin_cell.x] ) {
+                    const int distance = rl_dist( candidate.first, range->origin );
+                    if( distance > range->distance ) {
+                        continue;
+                    }
+                    sorted_elements[distance].push_back( candidate );
+                }
+            } else {
+                int min_cell_x = range->origin_cell.x - cell_ranks_loaded;
+                int min_cell_y = range->origin_cell.y - cell_ranks_loaded;
+                int max_cell_x = range->origin_cell.x + cell_ranks_loaded;
+                int max_cell_y = range->origin_cell.y + cell_ranks_loaded;
+                bool min_x_inbounds = min_cell_x > 0;
+                bool min_y_inbounds = min_cell_y > 0;
+                bool max_x_inbounds = max_cell_x <= MAXX / CELL_SIZE;
```

Comment:
If you do switch to `divide_round_up` then all these uses need to change too.  Could make a `static constexpr point cell_dimensions()` member function in `cell_list`.